### PR TITLE
Implement product clone

### LINKED ISSUE #11067: [BUU] As an enterprise user, I want to clone a product, so that I can add new items to my catalogue quickly

### PR DESCRIPTION
**Pull request by chahmedejaz**
#### What? Why?

- Closes #11067

#### What should we test?

- Validate all the scenarios mentioned in the issue above
- Also confirm that cloning in the old Bulk Product Edit screen still works as expected.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/views/admin/products_v3/components/_product_actions.html.haml
Comment (dacook) [inline app/views/admin/products_v3/components/_product_actions.html.haml:8]: I can see that this partial is used for two cases: product and variant (there are two types of record that can appear on this screen). 
This new action is only relevant for product, so it shouldn't show for a variant. I think this can be easily achieved simply by indenting to include in the above `else` block:

```suggestion
      = link_to t('admin.products_page.actions.clone'), clone_admin_product_path(product), class: "vertical-ellipsis-menu-content-item"
```

File: app/controllers/spree/admin/products_controller.rb
Comment (dacook) [inline app/controllers/spree/admin/products_controller.rb:92]: Although this looks like a change in behaviour, I think the frontend JS on the old bulk edit screen was reloading the products list anyway. 👍 
We'll test the old screen just in case though.

**Review by dacook — COMMENTED** [review]
Great work, and I have a suggestion to improve.

**Comment by chahmedejaz** [comment]
Thanks for the suggestions @dacook. Will incorporate those :)

File: app/controllers/spree/admin/products_controller.rb
Comment (chahmedejaz) [reply app/controllers/spree/admin/products_controller.rb:92]: Yes, you are right. This is a slight change in behavior as compared to the old screen.
We have [another issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/11737) which deals with actual behavior for the new screen.
This can be handled/updated in that issue I believe. 
Please let me know if you think otherwise. :)

File: app/controllers/spree/admin/products_controller.rb
Comment (chahmedejaz) [inline app/controllers/spree/admin/products_controller.rb:87]: Removing this to match with old UI experience.

Comment (dacook) [reply app/controllers/spree/admin/products_controller.rb:87]: Good idea to remove the unused code 👍 
Of course, we will try to implement these messages in the new UI, but the method will depend on https://github.com/openfoodfoundation/openfoodnetwork/issues/11517 which is in progress.

**Review by dacook — COMMENTED** [review]
Looks good so far, will await for Mario's comment

File: spec/system/admin/products_v3/products_spec.rb
Comment (dacook) [inline spec/system/admin/products_v3/products_spec.rb:367]: I know it's best practice to specify each expectation separately, but it can take a lot of resources to set up the test case each time. As these two don't affect each other, I would suggest putting them together to save time.

```suggestion
      it "shows an actions menu with a clone link when clicking on icon for product" do
        within row_containing_name("Apples") do
          page.find(".vertical-ellipsis-menu").click
          expect(page).to have_link "Clone", href: spree.clone_admin_product_path(product_a)
        end

        within row_containing_name("Medium box") do
          page.find(".vertical-ellipsis-menu").click
          expect(page).to_not have_link "Clone", href: spree.clone_admin_product_path(product_a)
        end
      end
```

**Review by dacook — COMMENTED** [review]
Great! I checked the specs and they look good too, although I have a suggestion to speed one up so will go ahead and try it.

Additionally, I've noticed that you've merged master into the branch a few times. Most of the time, it's ok when it says "This branch is out of date", but if there is a conflict, we do need to update it sometimes. 
So when we do this, our preference is to rebase in order to keep commit history clean. Could you please try that next time?

**Comment by dacook** [comment]
Looks like CI failures are due to rate limits: 
```
GET https://api.github.com/repos/openfoodfoundation/openfoodnetwork/pulls/11785: 403 API rate limit exceeded for installation ID 12364678. If you reach out to GitHub Support for help, please include the request ID 4F00:16E4:8C6FF8:91229A:6556A46C. [rate reset in 18m37s]
```

Will have to try later..

**Review by dacook — APPROVED** [review]

**Comment by dacook** [comment]
Awaiting a second review and manual testing.

File: spec/system/admin/products_v3/products_spec.rb
Comment (mkllnk) [inline spec/system/admin/products_v3/products_spec.rb:366]: I find it unnecessary to have system tests to check the display of an action button. You don't care if a button is displayed unless you want to click on it. So a test that clicks on a button is the important one and it checks the existence at the same time.
```suggestion
```

System tests are quite expensive to run and with many contributors our tests are often run dozens of times per hour. We also need to consider the environmental impact.

**Review by mkllnk — APPROVED** [review]
Great thank you!

I will adjust one little thing and rebase at the same time.

File: spec/system/admin/products_v3/products_spec.rb
Comment (mkllnk) [reply spec/system/admin/products_v3/products_spec.rb:366]: Actually, it does make sense here to check that the same button is in both places, variant and product. It could be further combined but I'll just leave it for now.

Comment (dacook) [reply spec/system/admin/products_v3/products_spec.rb:366]: I was thinking that conditions in view files should be tested with a view spec, which is way more efficient.

**Comment by chahmedejaz** [comment]
Thank you @dacook for the suggestions. I'll take care of the rebase thing next time. 😄

**Comment by RachL** [comment]
Seems to work as expected, thanks @chahmedejaz :muscle: 

Merging!

**Pull request merged**
